### DISCUSSION TITLE
[4.0] Joomla.asset.json

### DISCUSSION THREAD

**Discussion by brianteeman** [joomla/joomla-cms#32408] [Q&A]
1. If you use the copy template feature for the cassiopeia template all the references to cassiopeia in the joomla.asset.json file are maintained.
- Does that matter?
- Shouldnt they be renamed to the same value as the one you input when you copy the template?

2. Is there any documentation explaining how to use this file? What its purpose is etc etc. All I can find is https://docs.joomla.org/J4.x:Web_Assets which is way over my head. From what I have been able to deduce there are various joomla.asset.json all of which are loaded before the one for the template. So I am guessing that anything in the template joomla.asset.json file is either specific to that template and/or an override for the previous files.

I found a bit more information in the json-schema https://developer.joomla.org/schemas/json-schema/web_assets.json

If its technically possible to add comments to the joomla.asset.json file then it would be really helpful for anyone looking at cassiopeia as an example to build their new template with.

PS so far building a new site in j4 with bs5 has been a dream and lighthouse loves it. 99/100/100/100

**Comment by dgrammatiko** [comment]
> If you use the copy template feature for the cassiopeia template all the references to cassiopeia in the joomla.asset.json file are maintained.

That should be fine, we only have to move the cassiopeia and atum to the new template mode (the one that supports child templates). I have to finish the work here https://github.com/joomla/joomla-cms/pull/30359 In case that the templates finally ship with the legacy mode and someone forks the template then the  file ` joomla.asset.json ` needs to be adjusted

> PS so far building a new site in j4 with bs5 has been a dream and lighthouse loves it. 99/100/100/100

you can get 100 across the board with https://github.com/joomla/joomla-cms/pull/32141 😉

**Comment by Fedik** [comment]
> Does that matter?

Yes, much matter. Because it used by the "template", in various layouts:
https://github.com/joomla/joomla-cms/blob/c0a7b03c2c147e3a7e017b5e5428cbd71f163993/templates/cassiopeia/index.php#L56-L62
https://github.com/joomla/joomla-cms/blob/c0a7b03c2c147e3a7e017b5e5428cbd71f163993/templates/cassiopeia/error.php#L55-L60

etc.

> Shouldnt they be renamed to the same value as the one you input when you copy the template?

No, because it will break the template.
However, if you like, you can rename it manually with all entries in the template files.

> Is there any documentation explaining how to use this file?
> All I can find is https://docs.joomla.org/J4.x:Web_Assets 

Yes, you found correct, it is documentation :smile: 
Well, if you have an idea how it can be explained more understandable, that would be nice. Me and @wilsonge  wrote all what we can :smile: 
(Perhaps there need some update about recent changes, but in general it is not much related to your question.)

**Reply by brianteeman** [reply]
> > Shouldnt they be renamed to the same value as the one you input when you copy the template?

> No, because it will break the template.
> However, if you like, you can rename it manually with all entries in the template files.

I dont follow you on this. I create a copy of cassiopeia and give it the name "starter". The templateDetails.xml is updated to say "starter" and the folder is called "starter". So why cant we update the assets.json file that is copied into the new template as well and instead have to manually update it.

> Well, if you have an idea how it can be explained more understandable, that would be nice. Me and @wilsonge wrote all what we can 😄

It is very developer focused and not friendly to someone at my skill level. If you look at configuration.php-dist various lines are documented which is a start
https://github.com/joomla/joomla-cms/blob/c0a7b03c2c147e3a7e017b5e5428cbd71f163993/installation/configuration.php-dist#L64-L68

I have worked out what some of the entries are supposed to be and changed accordingly. But others I dont have a clue what they mean.

https://github.com/joomla/joomla-cms/blob/c0a7b03c2c147e3a7e017b5e5428cbd71f163993/templates/cassiopeia/joomla.asset.json#L30-L36

**Reply by Fedik** [reply]
> I dont follow you on this. I create a copy of cassiopeia and give it the name "starter". 

Look at the code I have shown previously, where it used in the template. Example index.php : 
https://github.com/joomla/joomla-cms/blob/c0a7b03c2c147e3a7e017b5e5428cbd71f163993/templates/cassiopeia/index.php#L56-L62

You cannot rename only in `json`. You need to rename all usages.
For test: try to rename in `json` manually in your copy, and will see what will happen.


> But others I dont have a clue what they mean.

`template.active.language`  load the css provided with the language (if there any)
`template.user` load user.css (style) and user.js (script)

> It is very developer focused and not friendly to someone at my skill level. 

Yeah. I think would be nice to add some extra explanation to the documentation for the "template how to" (if there any) with cross reference to Web_Assets.

> If its technically possible to add comments to the joomla.asset.json 

Yes it possible with `description` property, like it done for `template.active` asset https://github.com/joomla/joomla-cms/blob/c0a7b03c2c147e3a7e017b5e5428cbd71f163993/build/media_source/system/joomla.asset.json#L26-L31

**Reply by brianteeman** [reply]
> You cannot rename only in json. You need to rename all usages.

We can/should rename in all places - otherwise you can end up with someone uninstalling the original template as they think its not in use and it breaks the copied template which was in use

I have made a first draft at adding descriptions to the json file #32411 

> Yeah. I think would be nice to add some extra explanation to the documentation for the "template how to" (if there any) with cross reference to Web_Assets.

That is what I am really missing

**Reply by Fedik** [reply]
> We can/should rename in all places - otherwise you can end up with someone uninstalling the original template

No, we should not. It does not break if someone uninstalling the original template.
Because all assets copied also, and now they part of copied template.

`templates/cassiopeia/joomla.asset.json` refers to files within **cassiopeia** template
`templates/cassiopeia_copy/joomla.asset.json` refers to files within **cassiopeia_copy** template (even if there same names), previous file is ignored here.

Example an asset `template.cassiopeia.ltr` will look for `template.min.css` for active template. It would break, if URI was hardcoded to `templates/old_template/template.min.css` (It the same behavior as `JHtml::_('script', 'template.min.css')`)
https://github.com/joomla/joomla-cms/blob/421c3a15fd891c13559ea3f3ac7f5274aad6222f/templates/cassiopeia/joomla.asset.json#L8-L12


`joomla.asset.json` files of the template loaded only for active template, other templates is ignored.


Check screen form the docs:
![Screenshot_2021-02-14_11-28-39](https://user-images.githubusercontent.com/1568198/107873031-e67db280-6eb7-11eb-94b1-a55f82ed0dce.png)

That means only `system`, `vendor`, `legacy` `joomla.asset.json` always loaded. But component and template `joomla.asset.json`  only per active component and template.

**Comment by brianteeman** [comment]
I give up. I can't get bootstrap js loading in my custom template. Looking at cassiopeia as an example I cant see anything there that is loading bs. What am I missing

**Reply by dgrammatiko** [reply]
Give me an example of what interactive components you're trying to load. In core, the interactive components are loaded per case. So if you need the collapse you need to call `HTMLHelper::_('bootstrap.collapse', '.someCSSSelector');`, same pattern for anything else. `HTMLHelper::_('bootstrap.framework');` is loading (without initialising anything) all the components and shouldn't be used, like NEVER.

**Reply by brianteeman** [reply]
Simple case using just the default menu and the dropdown mobile menu isnt opening on my template

This is the same site just different templates
cassiopeia is showing this
![image](https://user-images.githubusercontent.com/1296369/108075845-bc1c2880-7062-11eb-862e-fe06a8b73ab6.png)


mytemplate is showing this
![image](https://user-images.githubusercontent.com/1296369/108076130-056c7800-7063-11eb-9d1f-5128dfa62057.png)

**Reply by dgrammatiko** [reply]
Check https://github.com/joomla/joomla-cms/blob/46216f83e16ea193825e478cbda9c7f793801be2/administrator/templates/atum/index.php#L129 This line tells Joomla to load the collapse component and initialise it for the `'.toggler-burger'` Css selector. 
Also https://github.com/joomla/joomla-cms/blob/46216f83e16ea193825e478cbda9c7f793801be2/administrator/modules/mod_user/tmpl/default.php#L19 is instructing Joomla to load the dropdown and initialise it for the `'.dropdown-toggle'` Css selector

**Reply by brianteeman** [reply]
Thanks for the hint.

Cassiopeia has this line in the template and I didnt

https://github.com/joomla/joomla-cms/blob/46216f83e16ea193825e478cbda9c7f793801be2/templates/cassiopeia/index.php#L132

**Reply by brianteeman** [reply]
I was so busy looking at the head and the assets I forgot to look in the body

**Reply by dgrammatiko** [reply]
I hope someone will pick up the Bootstrap demo component https://github.com/dgrammatiko/bs5/tree/main/test_component and polish it with proper documentation. It will be an invaluable help for newcomers

**Comment by brianteeman** [comment]
When you register an asset in your template joomla.asset.json there is a URI for each asset
But how does it know where these files live in the filesystem?

https://github.com/joomla/joomla-cms/blob/7c42442e8387cb2aa1b4852d65f2349bbb0d7298/templates/cassiopeia/joomla.asset.json#L12

https://github.com/joomla/joomla-cms/blob/7c42442e8387cb2aa1b4852d65f2349bbb0d7298/templates/cassiopeia/joomla.asset.json#L56

In other words if i want to add a new asset to my json where do I put the files in the template folder structure. I didnt see anything in the docs for this

**Reply by dgrammatiko** [reply]
It's the old convention: `.css` in the folder `css`, `.js` in the folder `js` and images in the images folder

**Reply by brianteeman** [reply]
So if I am adding my own web component to the template I need to split the js and css into those folders?

**Reply by brianteeman** [reply]
or perhaps I should/can do
uri: custom/js/file.js
uri: custom/css/file.css

with the files in /templatename/custom/css and /templatename/custom/js

**Reply by dgrammatiko** [reply]
You can do both, in reality, Web Assets still use the HTMLHelper to resolve the paths: https://github.com/joomla/joomla-cms/blob/338800d3ba8224dce10404aca8e791655101efc0/libraries/src/WebAsset/WebAssetItem.php#L300-L324

The difference is if the path is relative or not (~~I think you have to specify `'relative' => false` to use your own folder, the default is relative~~)

Basically if you want to use ` /templatename/custom/css` you have to set the uri to `templatename/custom/css/somefile.css` and it should work

**Reply by brianteeman** [reply]
ok - thanks

**Reply by Fedik** [reply]
> Basically if you want to use /templatename/custom/css you have to set the uri to templatename/custom/css/somefile.css and it should work

not really, it enough just `somefile.css`. Look how it in core templates. With this the AssetManager will pick `somefile.css` from active template. If template called "blabla" then in result the system will pick /templates/blabla/css/somefile.css.

**Reply by brianteeman** [reply]
Yes I tried a lot of combinations and there doesnt seem to be a way to load a script from any folder in the template other than one called js

**Reply by Fedik** [reply]
> ... other than one called js

~hm, I not very understood what you want to do,~
~what exactly you want to load?~

ah well, you do not want to use templates/my_template/js/ ? 
then you have to use full path to your script, as @dgrammatiko suggested, 
if your script in `templates/my_template/foobar/file.js` then you have to use this full path as `uri`

But then better to place your "web component" in to `/media` folder, `/media/foo-bar-fancy-stuff`, 
and use full path in `uri`: eg `media/foo-bar-fancy-stuff/path/to/foobar.js` 

Path resolving works like in joomla 3 with `HTMLHelper`

**Reply by brianteeman** [reply]
just to confirm I finally got it working with `"uri": "templates/templatename/video/script.js"`

**Reply by level420** [reply]
@brianteeman I could not completely follow all hints in this thread, but also like to load additional js files within my template which is a child template of `cassiopeia` called `cassiopeia_mysite`.

I've created a `joomla.asset.json` file in the directory `template/cassiopeia_mysite` file with the following content:
```json
{
  "$schema": "https://developer.joomla.org/schemas/json-schema/web_assets.json",
  "name": "cassiopeia_mysite",
  "version": "4.0.0",
  "description": "an additional js file which should be loaded",
  "license": "GPL-2.0-or-later",
  "assets": [
    {
      "name": "template.test",
      "description": "The name of a file where a user can add their own javascript",
      "type": "script",
      "uri":  "media/templates/site/cassiopeia_mysite/js/test.js"
    }
  ]
}
```
and of course the file `test.js` in `media/templates/site/cassiopeia_mysite/js`.

And honestly I don't know from the discussion if this should work or not, but it doesn't.

**Reply by dgrammatiko** [reply]
> but it doesn't.

It's expected if your `joomla.asset.json` looks like what you posted above. The explanation is that child templates inherit functionality from the parent template as is unless there is an override then they completely disregard the parent and use the local functionality in the child template. So, basically what it happens in the case of `joomla.asset.json` is:
- if the child template doesn't have the file `joomla.asset.json` the parent template's one is used
- if there is a file `joomla.asset.json` then that file should resolve **ALL** the static assets, there is no merging functionality. 

In short, if you want to use a file `joomla.asset.json` in a child template you have to redefine **all** the assets (most will have to point to the parent template **using absolute paths**, your extra entries could be relative to the child assuming the files are also under the child media place). I hope this is easy to follow

**Reply by level420** [reply]
@dgrammatiko thank you for your answer.

But in the setup which I've described above, it seems that the childs `joomla.asset.json` is ignored, as all assets which are defined in the parents `template/cassiopeia/joomla.asset.json` file still get loaded.

Anyway: if the assets defined in the child template do not get additionally loaded, this scenario is useless for me.

What I really need is a simple way to load additional javascript in a child template besides `user.js`. I'm not finding a general hook where I can do that, as all php overrides are specific to a layout, component, module etc. and therefore are only involved where those overrides are used. I'll need to load the additional files everywhere on the website.

**Reply by level420** [reply]
BTW.: I'm on Joomla 4.4.1

**Reply by Fedik** [reply]
In `template/foo-bar//joomla.asset.json` you defining template assets, they not loaded automaticlay.
They need to be enabled when needed in your template with `$wa->useScript('template.test')`.

Here is an example form Cassiopeia index.php, how it enable own assets: https://github.com/joomla/joomla-cms/blob/c0a7b03c2c147e3a7e017b5e5428cbd71f163993/templates/cassiopeia/index.php#L56-L62 

You can find more on Doc page https://docs.joomla.org/J4.x:Web_Assets
